### PR TITLE
Implement #2475

### PR DESCRIPTION
**Pull request by noahmayr**
Implementation for #2475

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/commands/log.rs
Comment (noahmayr) [inline cli/src/commands/log.rs:117]: Not sure if this part is even necessary at all, but I think it might be helpful to print out the revset that was used when no changes were found. If this is applicable to other commands, it might make sense to implement Display for RevsetExpression which would make it very easy to print out the used revset whether it was created from user input or programatically.

File: cli/src/commands/log.rs
Comment (yuja) [inline cli/src/commands/log.rs:98]: Perhaps, this kind of stuff can be done at the bottom to minimize the overhead.

**Review by yuja — COMMENTED** [review]
Regarding #2475, I think the default revset can be turned off if any path arguments are specified.
```
jj log            # use default revset
jj log -rREV      # REV
jj log PATH       # file(PATH)
jj log -rREV PATH # REV & file(PATH)
```

It implies that the `PATH` and `REV` arguments are basically the same things (with a bit different representations.) WDYT?

**Comment by joyously** [comment]
> Regarding # 2475, I think the default revset can be turned off if any path arguments are specified.
> 
> ```
> jj log            # use default revset
> jj log -rREV      # REV
> jj log PATH       # file(PATH)
> jj log -rREV PATH # REV & file(PATH)
> ```

What is the special case? If I do
```
jj log -r 'author(yuya)'
jj log CHANGELOG.md
jj log -r 'author(yuya)' CHANGELOG.md
```
I should get 3 different results.

**Comment by yuja** [comment]
> > ```
> > jj log            # use default revset
> > jj log -rREV      # REV
> > jj log PATH       # file(PATH)
> > jj log -rREV PATH # REV & file(PATH)
> > ```
> 
> What is the special case?

Special case exists only for `jj log` (without filtering options)?

The current definition is: `-r` defaults to `revsets.log`
So `jj log PATH` means `jj log -r<revsets.log> PATH`.

My proposal is: `revsets.log` is used only if no filtering options (`-r` or `PATH`) is specified.

**Comment by joyously** [comment]
> My proposal is: `revsets.log` is used only if no filtering options (`-r` or `PATH`) are specified.

Without knowing the code, it seems that the revset is needed (as in the bare `jj log`). 
If the default can't be set in the config file, then I guess you don't need it when there is a path.

**Comment by martinvonz** [comment]
I agree that it seems better to not use the default revset if any paths were passed. I think that's almost always what I want.

**Comment by noahmayr** [comment]
> Regarding #2475, I think the default revset can be turned off if any path arguments are specified.
> 
> ```
> jj log            # use default revset
> jj log -rREV      # REV
> jj log PATH       # file(PATH)
> jj log -rREV PATH # REV & file(PATH)
> ```
> 
> It implies that the `PATH` and `REV` arguments are basically the same things (with a bit different representations.) WDYT?

> I agree that it seems better to not use the default revset if any paths were passed. I think that's almost always what I want.

I added another commit which changes the behavior to reflect this. If it looks good to you I'll restructure the changes

File: cli/src/commands/log.rs
Comment (noahmayr) [reply cli/src/commands/log.rs]: If we want to automatically drop the revset and only filter for the path, then this needs to be at the top. Also it's only evaluating a second revset if both path and revset are specified and if the intersection of that is empty with the latest change

**Comment by yuja** [comment]
>        // if no args are passed, use default revset
>        // if only paths are passed, filter by paths
>        // if only a revset is passed, use the revset

These sound good to me, thanks.

>        // if paths and revset are passed, use both but try dropping the revset if the
>        // paths are not contained in it

Meh, no. If the user specified both, it's clear that they want to filter by both revset and paths.

BTW, I've changed `workspace_command.*revset()` APIs recently. We might need to make `attach_revset_evaluator()` public so the "only paths" case can be started from `RevsetExpresssion::all()`.

File: cli/src/commands/log.rs
Comment (yuja) [inline cli/src/commands/log.rs:41]: nit: "no paths nor revisions" for clarity?

File: cli/src/commands/log.rs
Comment (yuja) [inline cli/src/commands/log.rs:106]: Perhaps, the initial set can be encoded as follows:
```rust
if args.revisions.is_empty() && args.paths.is_empty() {
    workspace_command.parse_revset(&command.settings().default_revset())?
} else if !args.revisions.is_empty() {
    workspace_command.parse_union_revsets(&args.revisions)?
} else {
    workspace_command.attach_revset_evaluator(RevsetExpression::all())?
}
```

It will be probably easier to extend when implementing "diff" predicate.
https://github.com/martinvonz/jj/issues/2933#issuecomment-1925870731

**Review by yuja — APPROVED** [review]
LGTM, thanks. Can you squash these patches? We usually include doc/tests changes in the commit which introduce the new behavior.

File: cli/src/commands/log.rs
Comment (noahmayr) [reply cli/src/commands/log.rs]: agree, that sounds better

File: cli/src/commands/log.rs
Comment (noahmayr) [reply cli/src/commands/log.rs]: Updated and squashed

File: cli/src/commands/log.rs
Comment (yuja) [inline cli/src/commands/log.rs:83]: nit: this comment is obvious from the code, so let's remove it.

**Review by yuja — APPROVED** [review]
thanks!

File: cli/src/commands/log.rs
Comment (noahmayr) [reply cli/src/commands/log.rs:83]: oops actually wanted to do that since I added the comments for the specific statements 👍🏻

File: cli/src/commands/log.rs
Comment (ilyagr) [inline cli/src/commands/log.rs:43]: I would reword this slightly:

> Which revisions to show. If neither revisions nor paths are specified, the set of revisions defaults is the `revset.log` setting (if set) or `@ |ancestors(immutable_heads().., 2) | trunk()`. If any paths are specified, the default is `all()`. 

I think it could be improved further -- I'd like the short help to be, well, shorter, but I haven't come up with a great way to word that. If we only leave the first sentence in the short help, it will be easy to think that the argument is required or misunderstand the default.

**Comment by ilyagr** [comment]
Also, thank you!

File: cli/src/commands/log.rs
Comment (yuja) [reply cli/src/commands/log.rs:43]: Perhaps, the detailed doc can be moved up to the command level? Something like
* `jj log` by default ...
* if `-r REVISION` or `PATH` is specified, all visible revisions will be searched...

Comment (noahmayr) [reply cli/src/commands/log.rs:43]: So we leave this part for now?

File: docs/config.md
Comment (noahmayr) [reply docs/config.md:218]: Done

File: cli/src/commands/log.rs
Comment (yuja) [reply cli/src/commands/log.rs:43]: Yeah, I would address it separately.

File: docs/config.md
Comment (martinvonz) [inline docs/config.md:218]: Missing "nor"

**Review by martinvonz — APPROVED** [review]
Thanks!

File: docs/config.md
Comment (noahmayr) [reply docs/config.md]: Fixed 🤦🏻

**Review by ilyagr — APPROVED** [review]
I don't think you needed to wait for my approval, but just in case.

**Comment by noahmayr** [comment]
> I don't think you needed to wait for my approval, but just in case.

I'm not allowed to merge though, but judging by the fact that my name just turned green on discord I assume someone is already working on that :)

**Comment by martinvonz** [comment]
> I'm not allowed to merge though, but judging by the fact that my name just turned green on discord I assume someone is already working on that :)

I actually thought I already had done that. Done now anyway.

**Pull request merged**
